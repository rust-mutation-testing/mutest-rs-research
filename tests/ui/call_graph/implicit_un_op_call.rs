//@ print-call-graph
//@ print-targets
//@ stdout
//@ stderr: empty

use std::ops::{Deref, DerefMut, Neg, Not};

struct ImplsDeref;

fn deref_impl() {}
impl Deref for ImplsDeref {
    type Target = Self;

    fn deref(&self) -> &Self::Target {
        deref_impl();
        self
    }
}

#[test]
fn test_implicit_deref_call() {
    let _ = *ImplsDeref;
}

fn deref_mut_impl() {}
impl DerefMut for ImplsDeref {
    fn deref_mut(&mut self) -> &mut Self::Target {
        deref_mut_impl();
        self
    }
}

#[test]
fn test_implicit_deref_mut_call() {
    *ImplsDeref = ImplsDeref;
}

struct ImplsNeg;

fn neg_impl() {}
impl Neg for ImplsNeg {
    type Output = Self;

    fn neg(self) -> Self::Output {
        neg_impl();
        self
    }
}

#[test]
fn test_implicit_neg_call() {
    let _ = -ImplsNeg;
}

struct ImplsNot;

fn not_impl() {}
impl Not for ImplsNot {
    type Output = Self;

    fn not(self) -> Self::Output {
        not_impl();
        self
    }
}

#[test]
fn test_implicit_not_call() {
    let _ = !ImplsNot;
}
