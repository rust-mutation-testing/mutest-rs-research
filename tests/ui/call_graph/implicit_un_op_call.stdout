
@@@ call graph @@@

entry points:

test test_implicit_deref_call
  -> <ImplsDeref as std::ops::Deref>::deref at tests/ui/call_graph/implicit_un_op_call.rs:14:5: 14:37 (#0)
test test_implicit_deref_mut_call
  -> <ImplsDeref as std::ops::DerefMut>::deref_mut at tests/ui/call_graph/implicit_un_op_call.rs:27:5: 27:49 (#0)
test test_implicit_neg_call
  -> <ImplsNeg as std::ops::Neg>::neg at tests/ui/call_graph/implicit_un_op_call.rs:44:5: 44:33 (#0)
test test_implicit_not_call
  -> <ImplsNot as std::ops::Not>::not at tests/ui/call_graph/implicit_un_op_call.rs:61:5: 61:33 (#0)

nested calls at distance 1:

<ImplsDeref as std::ops::Deref>::deref at tests/ui/call_graph/implicit_un_op_call.rs:14:5: 14:37 (#0)
  -> deref_impl at tests/ui/call_graph/implicit_un_op_call.rs:10:1: 10:16 (#0)
<ImplsDeref as std::ops::DerefMut>::deref_mut at tests/ui/call_graph/implicit_un_op_call.rs:27:5: 27:49 (#0)
  -> deref_mut_impl at tests/ui/call_graph/implicit_un_op_call.rs:25:1: 25:20 (#0)
<ImplsNeg as std::ops::Neg>::neg at tests/ui/call_graph/implicit_un_op_call.rs:44:5: 44:33 (#0)
  -> neg_impl at tests/ui/call_graph/implicit_un_op_call.rs:40:1: 40:14 (#0)
<ImplsNot as std::ops::Not>::not at tests/ui/call_graph/implicit_un_op_call.rs:61:5: 61:33 (#0)
  -> not_impl at tests/ui/call_graph/implicit_un_op_call.rs:57:1: 57:14 (#0)

@@@ targets @@@

tests -(1)-> deref_impl at tests/ui/call_graph/implicit_un_op_call.rs:10:1: 10:16 (#0)
  (1) test_implicit_deref_call

tests -(0)-> <ImplsDeref as std::ops::Deref>::deref at tests/ui/call_graph/implicit_un_op_call.rs:14:5: 14:37 (#0)
  (0) test_implicit_deref_call

tests -(1)-> deref_mut_impl at tests/ui/call_graph/implicit_un_op_call.rs:25:1: 25:20 (#0)
  (1) test_implicit_deref_mut_call

tests -(0)-> <ImplsDeref as std::ops::DerefMut>::deref_mut at tests/ui/call_graph/implicit_un_op_call.rs:27:5: 27:49 (#0)
  (0) test_implicit_deref_mut_call

tests -(1)-> neg_impl at tests/ui/call_graph/implicit_un_op_call.rs:40:1: 40:14 (#0)
  (1) test_implicit_neg_call

tests -(0)-> <ImplsNeg as std::ops::Neg>::neg at tests/ui/call_graph/implicit_un_op_call.rs:44:5: 44:33 (#0)
  (0) test_implicit_neg_call

tests -(1)-> not_impl at tests/ui/call_graph/implicit_un_op_call.rs:57:1: 57:14 (#0)
  (1) test_implicit_not_call

tests -(0)-> <ImplsNot as std::ops::Not>::not at tests/ui/call_graph/implicit_un_op_call.rs:61:5: 61:33 (#0)
  (0) test_implicit_not_call

targets: 8 total; 8 safe; 0 unsafe (0 tainted)
